subsys/bluetooth/host/rfcomm.c

```diff
@@ -50,6 +50,18 @@ LOG_MODULE_REGISTER(bt_rfcomm);
 #define SESSION_RTX(_w) CONTAINER_OF(k_work_delayable_from_work(_w), \
 				     struct bt_rfcomm_session, rtx_work)
 
+struct bt_rfcomm_tx {
+	sys_snode_t node;
+	struct bt_rfcomm_dlc *dlc;
+	struct net_buf *buf;
+};
```

Comment:
Is this supposed to be analogous to `struct bt_conn_tx`? The primary purpose of `bt_conn_tx` is to allow the freeing of the tx buffer early (so that it can be reused again). However, here you're actually retaining a pointer to the buffer, which would defeat that purpose. Or is this for some other purpose?

Reply:
In the current rfcomm implementation, the data is not sent after the previous completion. It means there may be more than one tx buffer being transferred for the same DLC at the same time. And I need to use TX buffer to identify which `bt_conn_tx_cb_t` belongs to which data sending request.

For the data sent callback, there are two options.
Option 1, change the data sending mechanism. Use a FIFO to queue all TX buffers. And send them one by one, while ensuring that only one buffer is transmitting at the same time. It is similar to the logic of TX queue of `conn.c`. In this way we do not need to save TX buffer, because if the data sent callback notified, the tx buffer is current sending buffer.

Option 2, Save DLC and TX buffer to `struct bt_rfcomm_tx` for each sending request, and pass the pointer of `struct bt_rfcomm_tx` as user_data to bt_l2cap_chan_send_cb. If the data sent callback notified, the TX buffer and DLC can be got from parameter user_data of callback.

Reply:
Maybe I'm misunderstanding something, but couldn't you use the user data feature of net_bufs to associate a net_buf with a DLC?

Reply:
There is a very interesting thing here. This `struct net_buf::user_data` is used in the `bt_conn_send_cb` function. One possible way is to set the size of `struct net_buf::user_data` to `CONFIG_BT_CONN_TX_USER_DATA_SIZE` + `CONFIG_BT_RFCOMM_TX_USER_DATA_SIZE`. CONN uses `struct net_buf::user_data[0~CONFIG_BT_CONN_TX_USER_DATA_SIZE-1]`, and RFCOMM uses t`struct net_buf::user_data[CONFIG_BT_CONN_TX_USER_DATA_SIZE ~ CONFIG_BT_CONN_TX_USER_DATA_SIZE` + `CONFIG_BT_RFCOMM_TX_USER_DATA_SIZE-1]`. This method brings tight coupling between RFCOMM and CONN using `struct net_buf::user_data`, which brings potential risks to maintenance and update.

Reply:
That's a good point, i.e. the stack may overwrite any user data that the higher layers use. Another alternative, if the buffers come from the same buffer pool, is to use `net_buf_id()`. That's normally combined with a separate array of user data with the same number of elements as you have buffers in the pool.

Reply:
Yes, we can know its pool by id. But we also need a data structure to establish the mapping between pool id/buffer and DLC, because it is possible for multiple DLCs to share the same pool. 

Another approach I have considered is to use the TX buffer destroy callback which can be used as a TX buffer send completion notification. The disadvantage of this method is that regardless of whether the data is sent correctly, the TX buffer will be destroyed, and it is not sure who destroyed it. Because the error code will not be sent to you when it is destroyed, it is difficult to determine whether the data is actually sent.

Reply:
Yeah, the destroy callback isn't suitable for this purpose. I'm fine with the current approach you're taking here, but let's try to see if we could simplify & optimize this in the future.